File: FPSMonitor/FPSMonitor/FPSMonitor.m
//
//  FPSMonitor.m
//  MiU
//
//  Created by laughing on 15/7/15.
//  Copyright (c) 2015年 MiU. All rights reserved.
//

#import "FPSMonitor.h"
#import <UIKit/UIKit.h>

@interface FPSMonitor()
@property (nonatomic) UILabel *label;
@property (nonatomic) CADisplayLink *disp;
+ (instancetype)defaultMonitor;
@end

@implementation FPSMonitor

+ (instancetype)defaultMonitor
{
    static FPSMonitor *obj;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        obj = [self new];
    });
    return obj;
}

- (void)tick:(CADisplayLink*)disp
{
    static int frames;
    static double timeStamp;
    double now = CFAbsoluteTimeGetCurrent();
    if (now-timeStamp>0.5)
    {
        int val = 1.0*disp.frameInterval*frames/(now-timeStamp);
        _label.text = [NSString stringWithFormat:@"FPS:%i", val];
        if (val>=45)
        {
            _label.textColor = [UIColor greenColor];
        }else if (val>=30)
        {
            _label.textColor = [UIColor orangeColor];
        }else
        {
            _label.textColor = [UIColor redColor];
        }
        timeStamp = now;
        frames = 0;
    }else
    {
        ++frames;
    }
}

- (void)showFPS
{
    if (_label.superview)
    {
        [[UIApplication sharedApplication].keyWindow bringSubviewToFront:_label];
    }else
    {
        _label = [[UILabel alloc] initWithFrame:CGRectMake(2, 64, 80, 20)];
        [[UIApplication sharedApplication].keyWindow addSubview:_label];
        _label.backgroundColor = [UIColor clearColor];
    }
    
    if (!_disp)
    {
        _disp = [CADisplayLink displayLinkWithTarget:self selector:@selector(tick:)];
        _disp.frameInterval = 2;
        [_disp addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
    }
}

- (void)hideFPS
{
    [_disp invalidate];
    _disp = nil;
    [_label removeFromSuperview];
    _label = nil;
}

+ (void)showFPS
{
    [[self defaultMonitor] showFPS];
}

+ (void)hideFPS
{
    [[self defaultMonitor] hideFPS];
}

@end

File: FPSMonitor/FPSMonitor/FPSMonitor.swift
//
//  FPSMonitor.swift
//  FPSMonitor
//
//  Created by WG on 2018/3/23.
//  Copyright © 2018年 WG. All rights reserved.
//

import UIKit

public class FPSMonitor: NSObject {
    public static func show(){
        `default`.show()
    }
    public static func hide(){
        `default`.hide()
    }
    
    fileprivate var label = UILabel(frame: CGRect(x: 2, y: 22, width: 80, height: 20))
    fileprivate var disp: CADisplayLink?
    fileprivate var timeStamp = Double(0)
    fileprivate var frames = 0
    fileprivate static let `default` = FPSMonitor()
    
    @objc fileprivate func refresh(_ disp: CADisplayLink){
        let now = CFAbsoluteTimeGetCurrent()
        if now - timeStamp > 0.2{
            let val = Double(disp.frameInterval * frames) / (now - timeStamp)
            label.text = "FPS:\(Int(val))"
            if val >= 45{
                label.textColor = .green
            }else if val >= 30{
                label.textColor = .orange
            }else{
                label.textColor = .red
            }
            timeStamp = now
            frames = 0;
        }else{
            frames += 1
        }
        print(CFAbsoluteTimeGetCurrent())
    }
    fileprivate func show(){
        if let _ = label.superview {
            label.superview?.bringSubview(toFront: label)
            label.backgroundColor = .clear
        }else{
            UIApplication.shared.keyWindow?.addSubview(label)
        }
        if disp == nil{
            disp = CADisplayLink(target: self, selector: #selector(refresh(_:)))
            disp?.add(to: .main, forMode: .commonModes)
        }
    }
    fileprivate func hide(){
        disp?.invalidate()
        disp = nil
        label.removeFromSuperview()
    }
}

